Fails to compile for me when I use the debug flag:
```
$ go build -tags=xnetdebug ./x/xnet/
x/xnet/xnet_debug_pcap.go:17:5: undefined: internal
x/xnet/xnet_debug_pcap.go:18:3: undefined: internal
```

In addition: There seems to be an issue currently with the `debugheaplog` buildflag. In the default case e.g. `!debugheaplog` there is a call in StackAsync.Debug and StackAsync.DebugErr to LogAllocs unconditionally which could reduce performance, no? I think there might also be some footguns with the logging, like the [ALLOC] lines going straight to stderr which bypasses slog and LogAllocs runs before LogAttrs so allocations made by the LogAttrs call get attributed to the next call instead. Overall just minor stuff to eventually look into apart from the compile error (I think it's just missing the import)

--- a/internal/debug.go
+++ b/internal/debug.go
@@ -2,16 +2,16 @@ package internal
 
 import (
 	"log/slog"
+	"os"
 	"runtime"
 	"strconv"
 	"sync"
-	"unsafe"
 )
 
 const (
 	LevelTrace slog.Level = slog.LevelDebug - 2
 
-	usePrintLogAllocs = true
+	usePrintLogAllocs = HeapAllocDebugging
 )
 
 var (
@@ -19,11 +19,11 @@ var (
 	lastAllocs  uint64
 	lastMallocs uint64
 	allocmu     sync.Mutex
-	allocbuf    [256]byte
+	allocbuf    [128]byte
 )
 
-func LogAttrsAndAllocs(l *slog.Logger, level slog.Level, msg string, attrs ...slog.Attr) {
-	logAttrsAndAllocs(l, level, msg, attrs...)
+func LogAttrsAndAllocs(allocmsg string, l *slog.Logger, level slog.Level, msg string, attrs ...slog.Attr) {
+	logAttrsAndAllocs(allocmsg, l, level, msg, attrs...)
 }
 
 func LogAllocs(msg string) {
@@ -33,23 +33,34 @@ func LogAllocs(msg string) {
 		allocmu.Unlock()
 		return
 	}
+	inc := int64(memstats.TotalAlloc) - int64(lastAllocs)
+	numAlloc := int64(memstats.TotalAlloc) - int64(lastAllocs)
+	free := memstats.HeapSys - memstats.HeapInuse
 	if usePrintLogAllocs {
+		// Branch used when debugheaplog enabled
 		print("[ALLOC] ", msg)
-		print(" inc=", int64(memstats.TotalAlloc)-int64(lastAllocs))
-		print(" n=", int64(memstats.Mallocs)-int64(lastMallocs))
+		print(" inc=", inc)
+		print(" n=", numAlloc)
 		print(" heap=", memstats.HeapAlloc)
-		print(" free=", memstats.HeapSys-memstats.HeapInuse)
+		print(" free=", free)
 		print(" tot=", memstats.TotalAlloc)
 		println()
 	} else {
 		n := copy(allocbuf[:], "[ALLOC] ")
 		n += copy(allocbuf[n:], msg)
-		n += copyValueInt(allocbuf[n:], "inc", int64(memstats.TotalAlloc)-int64(lastAllocs))
-		n += copyValueInt(allocbuf[n:], "n", int64(memstats.Mallocs)-int64(lastMallocs))
+		n += copyValueInt(allocbuf[n:], "inc", inc)
+		n += copyValueInt(allocbuf[n:], "n", numAlloc)
 		n += copyValueUint(allocbuf[n:], "heap", memstats.HeapAlloc)
-		n += copyValueUint(allocbuf[n:], "free", memstats.HeapSys-memstats.HeapInuse)
+		n += copyValueUint(allocbuf[n:], "free", free)
 		n += copyValueUint(allocbuf[n:], "tot", memstats.TotalAlloc)
-		println(unsafe.String(&allocbuf[0], n))
+		allocbuf[n] = '\n'
+		os.Stdout.Write(allocbuf[:n+1])
+		if n > len(allocbuf) {
+			n2 := copy(allocbuf[:], "[WARN] ALLOC BUF OVERRUN")
+			n2 += copyValueUint(allocbuf[n2:], "n", uint64(n))
+			allocbuf[n2] = '\n'
+			os.Stdout.Write(allocbuf[:n2+1])
+		}
 	}
 	lastAllocs = memstats.TotalAlloc
 	lastMallocs = memstats.Mallocs

--- a/internal/debug.go
+++ b/internal/debug.go
@@ -34,7 +34,7 @@ func LogAllocs(msg string) {
 		return
 	}
 	inc := int64(memstats.TotalAlloc) - int64(lastAllocs)
-	numAlloc := int64(memstats.TotalAlloc) - int64(lastAllocs)
+	numAlloc := int64(memstats.Mallocs) - int64(lastMallocs)
 	free := memstats.HeapSys - memstats.HeapInuse
 	if usePrintLogAllocs {
 		// Branch used when debugheaplog enabled
@@ -46,20 +46,20 @@ func LogAllocs(msg string) {
 		print(" tot=", memstats.TotalAlloc)
 		println()
 	} else {
-		n := copy(allocbuf[:], "[ALLOC] ")
-		n += copy(allocbuf[n:], msg)
-		n += copyValueInt(allocbuf[n:], "inc", inc)
-		n += copyValueInt(allocbuf[n:], "n", numAlloc)
-		n += copyValueUint(allocbuf[n:], "heap", memstats.HeapAlloc)
-		n += copyValueUint(allocbuf[n:], "free", free)
-		n += copyValueUint(allocbuf[n:], "tot", memstats.TotalAlloc)
-		allocbuf[n] = '\n'
+		buf := allocbuf[:len(allocbuf)-1] // keep one character for newline.
+		n := copy(buf[:], "[ALLOC] ")
+		n += copy(buf[n:], msg)
+		n += copyValueInt(buf[n:], "inc", inc)
+		n += copyValueInt(buf[n:], "n", numAlloc)
+		n += copyValueUint(buf[n:], "heap", memstats.HeapAlloc)
+		n += copyValueUint(buf[n:], "free", free)
+		lastN := copyValueUint(buf[n:], "tot", memstats.TotalAlloc)
+		n += lastN
+		allocbuf[n] = '\n' // n will never be larger than len(allocbuf)-1
 		os.Stdout.Write(allocbuf[:n+1])
-		if n > len(allocbuf) {
-			n2 := copy(allocbuf[:], "[WARN] ALLOC BUF OVERRUN")
-			n2 += copyValueUint(allocbuf[n2:], "n", uint64(n))
-			allocbuf[n2] = '\n'
-			os.Stdout.Write(allocbuf[:n2+1])
+		if lastN == 0 {
+			n2 := copy(allocbuf[:], "[WARN] ALLOC BUF OVERRUN\n")
+			os.Stdout.Write(allocbuf[:n2])
 		}
 	}
 	lastAllocs = memstats.TotalAlloc

--- a/internal/debug_heaplog.go
+++ b/internal/debug_heaplog.go
@@ -4,7 +4,6 @@ package internal
 
 import (
 	"log/slog"
-	"runtime"
 	"time"
 	"unsafe"
 )
@@ -22,14 +21,13 @@ func LogEnabled(l *slog.Logger, lvl slog.Level) bool {
 	return true
 }
 
-func logAttrsAndAllocs(l *slog.Logger, level slog.Level, msg string, attrs ...slog.Attr) {
+func logAttrsAndAllocs(allocmsg string, l *slog.Logger, level slog.Level, msg string, attrs ...slog.Attr) {
 	LogAttrs(nil, level, msg, attrs...) // already logs attributes
 }
 
 func LogAttrs(_ *slog.Logger, level slog.Level, msg string, attrs ...slog.Attr) {
 	now := time.Now()
 	n := len(now.AppendFormat(timebuf[:0], timefmt))
-	LogAllocs(msg)
 	print("time=", unsafe.String(&timebuf[0], n), " ")
 	if level == LevelTrace {
 		print("TRACE ")
@@ -39,7 +37,6 @@ func LogAttrs(_ *slog.Logger, level slog.Level, msg string, attrs ...slog.Attr) 
 		print(level.String(), " ")
 	}
 	print(msg)
-
 	for _, a := range attrs {
 		switch a.Value.Kind() {
 		case slog.KindString:
@@ -53,12 +50,5 @@ func LogAttrs(_ *slog.Logger, level slog.Level, msg string, attrs ...slog.Attr) 
 		}
 	}
 	println()
-	allocmu.Lock()
-	runtime.ReadMemStats(&memstats)
-	if lastAllocs != memstats.TotalAlloc {
-		print("alloc increase in heaplog")
-	}
-	lastAllocs = memstats.TotalAlloc
-	lastMallocs = memstats.Mallocs
-	allocmu.Unlock()
+	LogAllocs(msg)
 }

--- a/internal/debug_noheaplog.go
+++ b/internal/debug_noheaplog.go
@@ -22,7 +22,8 @@ func LogAttrs(l *slog.Logger, level slog.Level, msg string, attrs ...slog.Attr) 
 	}
 }
 
-func logAttrsAndAllocs(l *slog.Logger, level slog.Level, msg string, attrs ...slog.Attr) {
-	LogAllocs(msg)
+func logAttrsAndAllocs(allocmsg string, l *slog.Logger, level slog.Level, msg string, attrs ...slog.Attr) {
+	LogAllocs(allocmsg)
 	LogAttrs(l, level, msg, attrs...)
+	LogAllocs(msg) // Should not log again unless LogAttrs allocated.
 }

--- a/x/xnet/stack-async.go
+++ b/x/xnet/stack-async.go
@@ -844,7 +844,7 @@ func addr4(addr [4]byte, ok bool) netip.Addr {
 
 // Debug prints debugging and heap information.
 func (s *StackAsync) Debug(msg string) {
-	internal.LogAttrsAndAllocs(slog.Default(), slog.LevelDebug, "stackasync",
+	internal.LogAttrsAndAllocs(msg, slog.Default(), slog.LevelDebug, "stackasync",
 		slog.String("umsg", msg),
 		slog.Uint64("sent", s.stats.TotalSent),
 		slog.Uint64("recv", s.stats.TotalReceived),
@@ -853,7 +853,7 @@ func (s *StackAsync) Debug(msg string) {
 
 // DebugErr prints debugging and heap information.
 func (s *StackAsync) DebugErr(msg, err string) {
-	internal.LogAttrsAndAllocs(slog.Default(), slog.LevelError, "stackasync",
+	internal.LogAttrsAndAllocs(msg, slog.Default(), slog.LevelError, "stackasync",
 		slog.String("umsg", msg),
 		slog.String("err", err),
 		slog.Uint64("sent", s.stats.TotalSent),

--- a/x/xnet/xnet_debug_pcap.go
+++ b/x/xnet/xnet_debug_pcap.go
@@ -2,7 +2,11 @@
 
 package xnet
 
-import "os"
+import (
+	"os"
+
+	"github.com/soypat/lneto/internal"
+)
 
 var _pcap CapturePrinter
 

